### PR TITLE
enh(cpp) Improve C++ class template headers

### LINKED ISSUE #2102: (cpp) grammar confused by complex Template << >> patterns
This code results in everything after `L':'` to be highlighted as if it was a string: **(Editor: This is now resolved, the second issue still standing).**

```cpp
template<typename T>
class PropertyChangedBase {
public:
    winrt::event_token PropertyChanged(const winrt::Windows::UI::Xaml::Data::PropertyChangedEventHandler &value)
    {
        return m_propertyChanged.add(value);
    }

    void PropertyChanged(const winrt::event_token &token)
    {
        m_propertyChanged.remove(token);
    }

protected:
    template<typename U>
    void compare_assign(U &value, const U &new_value, std::wstring_view name)
    {
        if (value != new_value)
        {
            value = new_value;
            name.remove_prefix(name.find_last_of(L':') + 1);
            m_propertyChanged(*static_cast<T *>(this), winrt::Windows::UI::Xaml::Data::PropertyChangedEventArgs(name));
        }
    }

private:
    winrt::event<winrt::Windows::UI::Xaml::Data::PropertyChangedEventHandler> m_propertyChanged;
};
```

On this one, it just completely stops highlighting after the second `is_streamable`

```cpp
namespace impl {
    template<typename T, typename = void>
    struct is_streamable : std::false_type { };

    template<typename T>
    struct is_streamable<T, std::void_t<decltype(std::declval<std::wostream &>() << std::declval<T>())>> : std::true_type { };

    template<typename T>
    inline constexpr bool is_streamable_v = is_streamable<T>::value;
}

// Disable overload for already valid operands.
template<class T, class = std::enable_if_t<!impl::is_streamable_v<const T &> && std::is_convertible_v<const T &, std::wstring_view>>>
std::wostream &operator <<(std::wostream &stream, const T &thing)
{
    return stream << static_cast<std::wstring_view>(thing);
}
```

Both pieces above is completely correct code which compiles, assuming the right headers are included

### LINKED ISSUE #2716: (cpp) Template parameters trigger class header rule
**Describe the issue**

When using the `template <class T>` syntax in C++, the `class` keyword misleads the parser into thinking it’s found a class header, and goes into class title mode, if there’s some `{` shortly afterwards. The relevant rule is

https://github.com/highlightjs/highlight.js/blob/5b87cc4d106059bee1c30f123e5a4053e0af9de4/src/languages/c-like.js#L227

**Which language seems to have the issue?**

`cpp`

**Are you using `highlight` or `highlightAuto`?**

`highlight`. When using `highlightAuto`, the code snippet is mis-categorised as `angelscript` and parsed differently still.

**Sample Code to Reproduce**

```cpp
template <class T> // comment
auto foo(T x) { … }
```

highlight.js marks this up as

```xml
<keyword>template</keyword> &lt;<class><keyword>class</keyword> <title>T</title>&gt; // <title>comment</title>
<title>auto</title> <title>foo</title>(<title>T</title> <title>x</title>) {</class> … }
```

**Expected behavior**

When replacing `class` with `typename` (which is 100% equivalent in this context), it’s marked up (more) correctly:

```xml
<keyword>template</keyword> &lt;<keyword>typename</keyword> T&gt; <comment>// comment</comment>
<function><keyword>auto</keyword> <title>foo</title><params>(T x)</params> </function>{ … }
```

**Additional context**

Present in master. Some earlier version used highlighted this correctly (presumably before support for class headers was added). Fundamentally the issue is that template parameter declarations aren’t handled at all by the lexer (by contrast, template argument lists *do* have rudimentary handling; i.e. usage rather than declaration).

A fix would probably require a separate mode for template parameter declarations, which is a hairy topic. In fact, correctly disambiguating the meaning of `>>` (can be right-shift operator *or* a terminating nested template argument list) is context-sensitive, and undecidable for the lexer (!) because its meaning can depend on the type of names it is applied to: `a<b<c>>d` can either be parsed as a variable declaration of `d` where the type has a nested template argument (list if `a` is a type), or as an expression involving less-than comparison and bit shifts, if `a` is a variable: https://godbolt.org/z/rM18oc

### PR DESCRIPTION
**Pull request by klmr**
This fixes #2716 and it fixes #2102<sup>1</sup>. In order to parse class and function headers correctly, it is insufficient to just skip greedily (or, worse, non-greedily) over `/<.*>/`, since C++ template declarations as well as template instantiations can contain nested template parameter lists. As a further complication, they can even contain the operators `<`, `>`, `<<` and `>>` (as part of non-type template parameters or their defaults (in declarations)).

This PR attempts to be a minimal change — a much more comprehensive refactor of the C and C++ language definitions is somewhat overdue (see #2173 and #2146). But this will take time, and in the meantime the improvement from this PR is probably “good enough”.

This PR is a WIP: there are some open issues that I’ve commented inside the changes.

---

<sup>1</sup> The failing code from the second issue isn’t added as a test case because hljs still fails to highlight it as a function declaration, but that’s unrelated to templates and rather because hljs currently doesn’t recognise operator overload definitions at all.

### CODE REVIEW

File: src/languages/c-like.js [hunk 1]
```diff
@@ -105,6 +105,7 @@ export default function(hljs) {
       'atomic_bool atomic_char atomic_schar ' +
       'atomic_uchar atomic_short atomic_ushort atomic_int atomic_uint atomic_long atomic_ulong atomic_llong ' +
       'atomic_ullong new throw return ' +
+      'class struct' +
```
Comment (klmr) [inline src/languages/c-like.js:108]: These keywords were missing generally: there are many situations (not just templates) in which they are valid and weren’t currently highlighted; `class/struct foo` is valid wherever `foo` is used as a type name, such as in declarations or in casts (`foo y = (class foo)(x)`).

File: src/languages/c-like.js [hunk 2]
```diff
@@ -200,6 +201,58 @@ export default function(hljs) {
     ]
   };
 
+  var TEMPLATE_USE = {
+    begin: /</, end: />/,
+    keywords: CPP_KEYWORDS,
+    contains: EXPRESSION_CONTAINS.concat([
+      // Match left-shift to prevent it from creating a nested `TEMPLATE_USE`.
```
Comment (klmr) [inline src/languages/c-like.js:208]: I’m repeating this quite a lot. I’m wondering whether this could be simplified, or whether the `/<</` mode should be moved inside `EXPRESSION_CONTAINS`.

File: src/languages/c-like.js [hunk 3]
```diff
@@ -200,6 +201,58 @@ export default function(hljs) {
...
+            contains: [
+              // Match left-shift to prevent it from creating a nested `TEMPLATE_USE`.
+              { begin: /<</ },
+              TEMPLATE_USE
+            ]
+          },
+        ]),
+        relevance: 10
+      },
+      // FIXME: Set relevance of class declarations at this point to 10!
```
Comment (klmr) [inline src/languages/c-like.js:251]: Can this be done without copying the definition of `CLASS_DECLARATION`? (It’s probably trivial but despite appearances to the contrary I don’t actually know JavaScript.)

File: test/markup/cpp/templates.txt [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+s.template foo<T>();
+class C;
+
+
+// Hard mode
+
+template <int a, int b, bool c = (a > b), class T> void f1();
+template <int a, int b, bool c = (a >> b), class T> void f2();
+
+template <int a, int b, bool c = (a < b), class T> void g1();
```
Comment (klmr) [inline test/markup/cpp/templates.txt:94]: This test is failing, and I don’t know why: I thought my rules would handle the nested parenthesised expression — and this works for the corresponding test case above, i.e. for `(a > b)`.

File: test/markup/cpp/templates.txt [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+class C;
+
+template <int a, int b, bool c = a < b, class T> void g1();
```
Comment (klmr) [inline test/markup/cpp/templates.txt:97]: This test is also failing, and there’s probably no fix, since it’s context sensitive: `a<b, class T>` *could* be a valid template parameter, if `a` were a template template parameter (i.e. if we had `template <typename> typename a` instead of `int a`). Of course in that case we’d be missing an additional terminating `>`, but I can’t see how to make the parser aware of this since it comes (potentially much) later.

**Review by klmr — COMMENTED** [review]
Request for comment:

File: test/markup/cpp/templates.txt (hunk 4 above)
Comment (joshgoebel) [reply test/markup/cpp/templates.txt:94]: I'd guess the `<` in `a < b` is starting a new TEMPLATE, messing everything up.

File: src/languages/c-like.js (hunk 2 above)
Comment (joshgoebel) [reply src/languages/c-like.js:208]: It doesn't look that terrible to me...

> whether the /<</ mode should be moved inside EXPRESSION_CONTAINS.

It doesn't technically belong there since it's only relevant in the context of a template, so I think the more explicit way it is now is slightly better than throwing it in there with a comment.

File: src/languages/c-like.js (hunk 3 above)
Comment (joshgoebel) [reply src/languages/c-like.js:251]: `inherit`

File: test/markup/cpp/templates.txt (hunk 5 above)
Comment (joshgoebel) [reply test/markup/cpp/templates.txt:97]: You could do someone like this with a callback that looks ahead and does code analysis, but that's definitely crossing over into "parsing" territory I think so I don't think we'd want to do that.

File: src/languages/c-like.js [hunk 6]
```diff
@@ -200,6 +201,58 @@ export default function(hljs) {
...
+
+  var TEMPLATE_DECLARATION = {
+    beginKeywords: 'template',
+    // Add additional stops to stop runaway template declaration, e.g.
+    // `template <bool b = 1 < 2> void f();`
+    end: /[>;{]/,
+    contains: [
+      {
+        begin: /</, end: />/,
+        endsWithParent: true,
```
Comment (joshgoebel) [inline src/languages/c-like.js:232]: Just so you know this `endsWithParent` does not nest... it simply means that within the scope of a `<...>` that it will attempt to match not on the contains rules on line 235, but ALSO the end rule on line 228.  But it doesn't go any deeper than that unless you added `endsWithParent` at every level...

I'm not sure if this has implications for your runaway guard on line 227 or not...

File: src/languages/c-like.js (hunk 1 above)
Comment (joshgoebel) [reply src/languages/c-like.js:108]: What often happens (and why these were excluded here) is false positives on `bob.class()`... does C++ allow class or struct as method names on classes?

File: test/markup/cpp/templates.txt [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+
+
+eval<A<int>> eA;
+eval<B<int, float>> eB;
+eval<C<17>> eC;
+eval<D<int, 17>> eD;
+eval<E<int, float>> eE;
+
+
+template <template <auto> class> void FA();  // note: C++17
```
Comment (joshgoebel) [inline test/markup/cpp/templates.txt:76]: This isn't a template match because it has a ` ` after template so the `<` will never match because the rule will end when it hits a space (which doesn't match any of your TEMPLATE rules...

Adding a space eater to your begin matcher for `<` should help.

Comment (joshgoebel) [reply test/markup/cpp/templates.txt:76]: Are you using the developer tool?  This kind of thing is a lot easier to see with the visual aids of block nesting.

File: src/languages/c-like.js [hunk 8]
```diff
@@ -200,6 +201,58 @@ export default function(hljs) {
     ]
   };
 
+  var TEMPLATE_USE = {
+    begin: /</, end: />/,
+    keywords: CPP_KEYWORDS,
+    contains: EXPRESSION_CONTAINS.concat([
+      // Match left-shift to prevent it from creating a nested `TEMPLATE_USE`.
+      { begin: /<</ },
+      'self'
```
Comment (joshgoebel) [inline src/languages/c-like.js:210]: This is going to get tripped up on `<` (less than) inside a template I think because you have no way to know if it's an operator or the start of a new TEMPLATE_USE, which is why your one test is failing I think.

Comment (joshgoebel) [reply src/languages/c-like.js:210]: Also can `<<` Truly never be a nested template use?

File: test/markup/cpp/templates.txt (hunk 7 above)
Comment (klmr) [reply test/markup/cpp/templates.txt:76]: Why would whitespace be an issue here? It works fine for `TEMPLATE_DECLARATION`. That said, nesting templates are actually an issue I had forgotten to comment on: how would I do that? I’ve got a circular dependency and I can’t use `self`, since the mode is nested deeper (i.e. `TEMPLATE_DECLARATION` does *not* contain `TEMPLATE_DECLARATION` — but it contains a mode which should contain `TEMPLATE_DECLARATION`).

In fact, I do use the developer tool, and here’s what I’m currently seeing for this input (I’ve added additional class names for debugging purposes):

Screenshot 2020-10-04 at 14 47 18

Ideally we’d want this to have another `tpl-decl` inside the `tpl-params`.

File: src/languages/c-like.js (hunk 8 above)
Comment (klmr) [reply src/languages/c-like.js:210]: Yes: each template declaration starts with `template`, and each template instantiation starts with an identifier, followed by a single `<`.

File: test/markup/cpp/templates.txt (hunk 4 above)
Comment (klmr) [reply test/markup/cpp/templates.txt:94]: Ah, of course! I actually don’t know how to handle this: I thought about adding `)` to the `end` of the `TEMPLATE_USE`, mode; this works, but then the mode started by `(` won’t terminate, even when I also set `excludeEnd: true`.

File: src/languages/c-like.js (hunk 1 above)
Comment (klmr) [reply src/languages/c-like.js:108]: It does not: C++ keywords are reserved in *all* contexts. There are some context-dependent reserved words (e.g. `override` and `final`) but these aren’t part of this list.

File: src/languages/c-like.js (hunk 6 above)
Comment (klmr) [reply src/languages/c-like.js:232]: Good question! So, if I understand you correctly, I could equivalently remove `endsWithParent, and change the `end` rule to `/[>;{]/` (i.e. the same as the parent `end`)? In this case I should probably do that instead, and add the same end to the other contained rules.

File: test/markup/cpp/templates.txt (hunk 4 above)
Comment (klmr) [reply test/markup/cpp/templates.txt:94]: Adding the following to the `TEMPLATE_USE` mode seems to work:

```javascript
    'on:begin': (matchData, response) => {
      const rest = matchData.input.slice(matchData.index + 1);
      let parens = 0;
      let brackets = 1; // start with `<`
      for (let i = 0; i < rest.length; i++) {
        const c = rest.charAt(i);
        switch (c) {
          case '(': parens++; break;
          case ')': parens--; break;
          case '<': brackets++; break;
          case '>': brackets--; break;
        }
        if (brackets == 0) return;
        if (parens == -1) {
          response.ignoreMatch();
          return;
        }
      }
    },
```

However, this feels like overkill (potentially also inefficient?); it also *only* works in cases where the comparison expression is nested inside parentheses, i.e. it works for this test case but not the next one.

If this approach is taken at all, we’d probably want to add some emergency stop characters (`;`, `{`) to the `switch`.

Comment (joshgoebel) [reply test/markup/cpp/templates.txt:94]: Yeah, this is too much... and definitely has performance implications.

File: src/languages/c-like.js (hunk 6 above)
Comment (joshgoebel) [reply src/languages/c-like.js:232]: Well if the desire is that they end with the parent you'd probably want to just use that... that's what it does - pull the parent "end" down into the children... I was just pointing out that it only goes one layer deep... so if it got "stuck" in a child of a child then your "guard" would not save you... so that needs to be considered and might result in adding some more "false positive" edge case tests.

File: test/markup/cpp/templates.txt (hunk 7 above)
Comment (joshgoebel) [reply test/markup/cpp/templates.txt:76]: Ah I see now, you don't have a class (which you shouldn't) - so I wasn't seeing it because I was only looking at the HTML. :-)  You have an `end` rule so it works... I missed.  If you didn't have and `end` the mode would immediately terminate if it could find no children.

> That said, nesting templates are actually an issue I had forgotten to comment on: how would I do that? I’ve got a circular dependency and I can’t use self, since the mode is nested deeper 

You'd have to do it manually (which will limit depth)... Or rejigger the rules so that say your template rule was say `template<` to `>`...  so that you could then use self...

File: test/markup/cpp/templates.txt (hunk 4 above)
Comment (joshgoebel) [reply test/markup/cpp/templates.txt:94]: I think the best maybe we can do here is liberal use of endsWithParent to make sure we escape if we hit a `{`}, line end, etc... and try to return to normal matching...

File: test/markup/cpp/templates.txt (hunk 7 above)
Comment (klmr) [reply test/markup/cpp/templates.txt:76]: The two nested modes are purely so that I could start with a (highlighted) keyword (`template`), followed by a non-keyword character. I initially had `begin: /template\s*</` but (I thought) that didn’t highlight `template` as a keyword.

But I can’t reproduce this now: it *does* highlight `template` correctly, so I guess I can simplify that nested mode and make it recursive.

File: src/languages/c-like.js (hunk 6 above)
Comment (klmr) [reply src/languages/c-like.js:236]: I think that works now. I needed to add a nested `{begin: />/}` to prevent a `>` inside a `(…)` from terminating the template declaration mode prematurely. That said, the behaviour on my unit tests is actually the same as before, I need to think of some example which works now and would previously have failed.

Comment (klmr) [reply src/languages/c-like.js:249]: Damn, I think I’ve just kicked the can down the road now. I think this fundamentally doesn’t work because I would need to nest the parent mode inside the mode that handles `(…)`. With my newest change, the the two first lines parse correctly, but the template declaration in the third line ends prematurely, and the “class title” starts at `class T` rather than at `class C`:

```cpp
template <int A = (B<C>), class T> class C;
template <int A = B<(c > D<E>)>, class T> class C;
template <int A = B<(c > D<(e > f)>)>, class T> class C;
```

File: test/markup/cpp/templates.txt (hunk 7 above)
Comment (joshgoebel) [reply test/markup/cpp/templates.txt:76]: > that didn’t highlight template as a keyword. ... But I can’t reproduce this now:

It should indeed IF it's in the `keyword` object at the same level... `beginKeywords` hooks this up automatically... if you're writing the rule by hand you'd need both `begin` and `keyword`.

File: src/languages/c-like.js (hunk 6 above)
Comment (joshgoebel) [reply src/languages/c-like.js:249]:
```
  var TEMPLATE_USE = {};
  Object.assign(TEMPLATE_USE, {
    // .. can now use TEMPLATE_USE recursively
```

All you need is a reference. :-)  Though I'm not sure we do this anywhere else and not sure it's a pattern we should start using now though. ...

I'm guessing the nesting here could be unlimited but is there some practical limits for MOST C++ code as far as what you'd see in the real world?

**Comment by joshgoebel** [comment]
Closing this in favor of #2752.  Thanks for all the hard work on this though - it's been appreciated!  Hopefully you've learned a bit about the grammars and parser that will help you if you continue to contribute. :-)

If we find more failure cases we can always re-review, but this was getting way to complicated.

**Comment by klmr** [comment]
Yes, I’m definitely in favour of closing this for now and re-submitting another PR later (in particular, *after* splitting the definitions for C and C++), I just need to find some time …
